// @wdio/local-runner/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _logger = _interopRequireDefault(require("@wdio/logger"));

var _streamBuffers = require("stream-buffers");

var _worker = _interopRequireDefault(require("./worker"));

var _constants = require("./constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

const log = (0, _logger.default)('@wdio/local-runner');

class LocalRunner {
  constructor(configFile, config) {
    this.configFile = configFile;
    this.config = config;
    this.workerPool = {};
    this.stdout = new _streamBuffers.WritableStreamBuffer(_constants.BUFFER_OPTIONS);
    this.stderr = new _streamBuffers.WritableStreamBuffer(_constants.BUFFER_OPTIONS);
  }

  initialise() {}

  getWorkerCount() {
    return Object.keys(this.workerPool).length;
  }

  run(_ref) {
    let {
      command,
      args
    } = _ref,
        workerOptions = _objectWithoutProperties(_ref, ["command", "args"]);

    const workerCnt = this.getWorkerCount();

    if (workerCnt >= process.stdout.getMaxListeners() - 2) {
      process.stdout.setMaxListeners(workerCnt + 2);
      process.stderr.setMaxListeners(workerCnt + 2);
    }

    const worker = new _worker.default(this.config, workerOptions, this.stdout, this.stderr);
    this.workerPool[workerOptions.cid] = worker;
    worker.postMessage(command, args);
    return worker;
  }

  shutdown() {
    log.info('Shutting down spawned worker');

    for (const [cid, worker] of Object.entries(this.workerPool)) {
      const {
        caps,
        server,
        sessionId,
        config,
        isMultiremote,
        instances
      } = worker;
      let payload = {};

      if (config && config.watch && (sessionId || isMultiremote)) {
        payload = {
          config: _objectSpread(_objectSpread({}, server), {}, {
            sessionId
          }),
          caps,
          watch: true,
          isMultiremote,
          instances
        };
      } else if (!worker.isBusy) {
        delete this.workerPool[cid];
        continue;
      }

      worker.postMessage('endSession', payload);
    }

    return new Promise(resolve => {
      const timeout = setTimeout(resolve, _constants.SHUTDOWN_TIMEOUT);
      const interval = setInterval(() => {
        const busyWorker = Object.entries(this.workerPool).filter(([, worker]) => worker.isBusy).length;
        log.info(`Waiting for ${busyWorker} to shut down gracefully`);

        if (busyWorker === 0) {
          clearTimeout(timeout);
          clearInterval(interval);
          log.info('shutting down');
          return resolve();
        }
      }, 250);
    });
  }

}

exports.default = LocalRunner;
