// @wdio/mocha-framework/build/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MOCHA_TIMEOUT_MESSAGE_REPLACEMENT = exports.MOCHA_TIMEOUT_MESSAGE = exports.NOOP = exports.EVENTS = exports.INTERFACES = void 0;
const INTERFACES = {
  bdd: ['it', 'before', 'beforeEach', 'after', 'afterEach'],
  tdd: ['test', 'suiteSetup', 'setup', 'suiteTeardown', 'teardown'],
  qunit: ['test', 'before', 'beforeEach', 'after', 'afterEach']
};
exports.INTERFACES = INTERFACES;
const EVENTS = {
  'suite': 'suite:start',
  'suite end': 'suite:end',
  'test': 'test:start',
  'test end': 'test:end',
  'hook': 'hook:start',
  'hook end': 'hook:end',
  'pass': 'test:pass',
  'fail': 'test:fail',
  'pending': 'test:pending'
};
exports.EVENTS = EVENTS;

const NOOP = function () {};

exports.NOOP = NOOP;
const MOCHA_TIMEOUT_MESSAGE = 'For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.';
exports.MOCHA_TIMEOUT_MESSAGE = MOCHA_TIMEOUT_MESSAGE;
const MOCHA_TIMEOUT_MESSAGE_REPLACEMENT = ['The execution in the test "%s %s" took too long. Try to reduce the run time or', 'increase your timeout for test specs (https://webdriver.io/docs/timeouts.html).'].join(' ');
exports.MOCHA_TIMEOUT_MESSAGE_REPLACEMENT = MOCHA_TIMEOUT_MESSAGE_REPLACEMENT;
