// local-runner/runner.js
const Hapi = require('hapi')

function makeRoute(server, method, route, handler) {
  console.log(`${server.info.uri}${route} ${method}`) // eslint-disable-line no-console
  return server.route({
    method: method,
    path: route,
    handler: function (request, reply) {
      return reply(handler(request))
    }
  })
}

module.exports = function () {
  const server = new Hapi.Server()
  server.connection({
    host: 'localhost',
    port: 8000
  })

  const blinkmrc = require(`${process.cwd()}/.blinkmrc.json`)

  blinkmrc.server.routes.map((route) => {

    const module = require(`${process.cwd()}/${route.module}`)

    if (typeof module === 'function') {
      // Add the route
      makeRoute(server, 'GET', route.route, module)
      makeRoute(server, 'POST', route.route, module)
      makeRoute(server, 'PUT', route.route, module)
      makeRoute(server, 'DELETE', route.route, module)
    } else {
      if (typeof module.get === 'function') {
        makeRoute(server, 'GET', route.route, module.get)
      }
      if (typeof module.post === 'function') {
        makeRoute(server, 'POST', route.route, module.post)
      }
      if (typeof module.put === 'function') {
        makeRoute(server, 'PUT', route.route, module.put)
      }
      if (typeof module.delete === 'function') {
        makeRoute(server, 'DELETE', route.route, module.delete)
      }
    }
  })

  // Start the server
  return server.start((err) => {

    if (err) {
      throw err
    }
    console.log('Server running at:', server.info.uri) // eslint-disable-line no-console
  })
}
