// puppeteer-core/lib/EventEmitter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventEmitter = void 0;
const mitt_1 = __importDefault(require("mitt"));
class EventEmitter {
    constructor() {
        this.listenerCounts = new Map();
        this.emitter = mitt_1.default(new Map());
    }
    on(event, handler) {
        this.emitter.on(event, handler);
        const existingCounts = this.listenerCounts.get(event);
        if (existingCounts) {
            this.listenerCounts.set(event, existingCounts + 1);
        }
        else {
            this.listenerCounts.set(event, 1);
        }
    }
    off(event, handler) {
        this.emitter.off(event, handler);
        const existingCounts = this.listenerCounts.get(event);
        this.listenerCounts.set(event, existingCounts - 1);
    }
    removeListener(event, handler) {
        this.off(event, handler);
    }
    emit(event, eventData) {
        this.emitter.emit(event, eventData);
    }
    once(event, handler) {
        const onceHandler = (eventData) => {
            handler(eventData);
            this.off(event, onceHandler);
        };
        this.on(event, onceHandler);
    }
    listenerCount(event) {
        return this.listenerCounts.get(event) || 0;
    }
}
exports.EventEmitter = EventEmitter;
