// @wdio/config/build/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SUPPORTED_FILE_EXTENSIONS = exports.SUPPORTED_HOOKS = exports.DEFAULT_CONFIGS = void 0;
const DEFAULT_TIMEOUT = 10000;

const DEFAULT_CONFIGS = () => ({
  specs: [],
  suites: {},
  exclude: [],
  outputDir: undefined,
  logLevel: 'info',
  logLevels: {},
  excludeDriverLogs: [],
  bail: 0,
  waitforInterval: 500,
  waitforTimeout: 5000,
  framework: 'mocha',
  reporters: [],
  services: [],
  maxInstances: 100,
  maxInstancesPerCapability: 100,
  filesToWatch: [],
  connectionRetryTimeout: 120000,
  connectionRetryCount: 3,
  execArgv: [],
  runnerEnv: {},
  runner: 'local',
  mochaOpts: {
    timeout: DEFAULT_TIMEOUT
  },
  jasmineNodeOpts: {
    defaultTimeoutInterval: DEFAULT_TIMEOUT
  },
  cucumberOpts: {
    timeout: DEFAULT_TIMEOUT
  },
  onPrepare: [],
  onWorkerStart: [],
  before: [],
  beforeSession: [],
  beforeSuite: [],
  beforeHook: [],
  beforeTest: [],
  beforeCommand: [],
  afterCommand: [],
  afterTest: [],
  afterHook: [],
  afterSuite: [],
  afterSession: [],
  after: [],
  onComplete: [],
  onReload: [],
  beforeFeature: [],
  beforeScenario: [],
  beforeStep: [],
  afterStep: [],
  afterScenario: [],
  afterFeature: []
});

exports.DEFAULT_CONFIGS = DEFAULT_CONFIGS;
const SUPPORTED_HOOKS = ['before', 'beforeSession', 'beforeSuite', 'beforeHook', 'beforeTest', 'beforeCommand', 'afterCommand', 'afterTest', 'afterHook', 'afterSuite', 'afterSession', 'after', 'beforeFeature', 'beforeScenario', 'beforeStep', 'afterStep', 'afterScenario', 'afterFeature', 'onReload', 'onPrepare', 'onWorkerStart', 'onComplete'];
exports.SUPPORTED_HOOKS = SUPPORTED_HOOKS;
const SUPPORTED_FILE_EXTENSIONS = ['.js', '.mjs', '.es6', '.ts', '.feature', '.coffee', '.cjs'];
exports.SUPPORTED_FILE_EXTENSIONS = SUPPORTED_FILE_EXTENSIONS;
